Feature/ar24 int8 optimizations

mpenkows opened the pull request:
CVS-34637

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/fused_conv_eltwise/fused_conv_eltwise_kernel_b_fs_yx_fsv4.cpp:
```diff
@@ -0,0 +1,109 @@
+    : fused_conv_eltwise_kernel_base("fused_conv_eltwise_gpu_b_fs_yx_fsv4") {
+}
+
+ParamsKey fused_conv_eltwise_kernel_b_fs_yx_fsv4::GetSupportedKey() const {
```
Eltwise Sum fusion can also be processed by FUSED_OPS codegen, thus fused_conv_eltwise_* kernels are deprecated and we plan to remove them eventually. Could you please move this kernel to common convolution folder and update accordingly?

mpenkows replied:
Fused_conv_eltwise kernel removed from this PR.

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_yx_fsv4_int8.cl:
```diff
@@ -0,0 +1,171 @@
+#define AS_INPUT0_TYPE_4(x) AS_TYPE_N(INPUT0_TYPE, 4, x)
+#define AS_FILTER_TYPE_4(x) AS_TYPE_N(FILTER_TYPE, 4, x)
+
+    const uint idx = 2 * ((uint)get_global_id(0) * 16 + (uint)get_global_id(2));
```
Please replace magic constants with something meaningful.

mpenkows replied:
Done.

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv4_int8.cpp:
```diff
@@ -0,0 +1,111 @@
+ParamsKey ConvolutionKernel_b_fs_yx_fsv4_int8::GetSupportedKey() const {
+    ParamsKey k;
+    k.EnableInputDataType(Datatype::F16);
+    k.EnableInputWeightsType(WeightsType::F16);
```
Does the kernel really support F16 input/weights? From what I see it works with (u)int8 only

mpenkows replied:
F16 support removed.

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv4_int8.cpp:
```diff
@@ -0,0 +1,111 @@
+
+    if (!params.fused_ops.empty()) {
+        auto input_dt = GetActivationType(params);
+        FusedOpsConfiguration conf0 = { "_0", {"batch", "FILTER_OFM_MAX * iter + ofm + 0", "idy", "idx"}, "res0", input_dt, 1 };
```
Do you really need these 8 different configurations? Can't fused ops be processed with single one + loop in the kernel?

mpenkows replied:
Wanted to avoid yet another loop in the kernel.

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_yx_fsv4_int8.cl:
```diff
@@ -0,0 +1,171 @@
+#    define ACTIVATION_TYPE float
+#    define TO_ACTIVATION_TYPE(x) convert_float(x)
+#else
+#    define ACCUMULATOR_TYPE INPUT0_TYPE
```
Accumulator/activation types should be defined on the host side.
```cpp
    jit.Merge(MakeTypeJitConstants(GetAccumulatorType(params), "ACCUMULATOR"));
    jit.Merge(MakeTypeJitConstants(GetActivationType(params), "ACTIVATION"));
```

mpenkows replied:
Done.

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_yx_fsv4_int8.cl:
```diff
@@ -0,0 +1,171 @@
+__attribute__((intel_reqd_sub_group_size(SUB_GROUP_SIZE)))
+__attribute__((reqd_work_group_size(1, 1, SUB_GROUP_SIZE)))
+KERNEL(convolution_gpu_b_fs_yx_fsv4_int8)(
+    const __global float* input,
```
float -> uint or define INPUT0_PACKED_TYPE and use it

mpenkows replied:
Done

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_yx_fsv4_int8.cl:
```diff
@@ -0,0 +1,171 @@
+            filter_idx = FILTER_OFM_MAX * iter * 4;
+
+            __attribute__((opencl_unroll_hint((FILTER_IFM_NUM + 3) / 4)))
+                for (int ifm = 0; ifm < (FILTER_IFM_NUM+3)/4; ifm++) {
```
Please fix code style in the kernel here and below
e.g. `(FILTER_IFM_NUM+3)/4` ->  `(FILTER_IFM_NUM + 3) / 4`

vladimir-paramuzov replied:
And fix alignment as well.

mpenkows replied:
Done

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/fused_conv_eltwise/fused_conv_eltwise_kernel_b_fs_yx_fsv4.cpp:
```diff
@@ -0,0 +1,109 @@
+ParamsKey fused_conv_eltwise_kernel_b_fs_yx_fsv4::GetSupportedKey() const {
+    ParamsKey k;
+    k.EnableInputDataType(Datatype::F16);
+    k.EnableInputDataType(Datatype::F32);
```
Same concern on supported data types as in the other kernel

mpenkows replied:
Fused_conv_eltwise kernel removed from this PR.

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/fused_conv_eltwise_gpu_b_fs_yx_fsv4.cl:
```diff
@@ -0,0 +1,210 @@
+
+__attribute__((intel_reqd_sub_group_size(SUB_GROUP_SIZE)))
+__attribute__((reqd_work_group_size(1, 1, SUB_GROUP_SIZE)))
+KERNEL(fused_conv_eltwise_kernel_b_fs_yx_fsv4)(
```
Is this kernel needed at all? Looks like you can just add the part that processes image2d output to the other conv fsv4 kernel, because the part that do the convolution looks pretty much the same

mpenkows replied:
Fused_conv_eltwise kernel removed from this PR.

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/tests/test_cases/convolution_gpu_test.cpp:
```diff
@@ -5353,7 +5353,99 @@ TEST_P(convolution_gpu_fs_byx_fsv32_crop, fs_byx_fsv32_crop)
                 }
 }
 
+TEST(convolution_f32_fw_gpu, convolution_int8_b_fs_yx_fsv4_to_bfyx) {
```
Test coverage should be extended for these kernels.

mpenkows replied:
Fused_conv_eltwise kernel removed from this PR so no test needed for it

vladimir-paramuzov replied:
Are there any test cases in fusings_gpu_tests with this kernel?

mpenkows commented:
I addressed all the remarks. I am also attaching a result from perf run. All tests passed. On our target model- AR24 we have almost 3 times better result.
[results.txt](https://github.com/openvinotoolkit/openvino/files/4999544/results.txt)

sshlyapn commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_b_fs_yx_fsv4_int8.cl:
```diff
@@ -0,0 +1,163 @@
+                        }
+                }
+
+            __attribute__((opencl_unroll_hint(FILTER_OFM_MAX / 4)))
+                for (int ofm = 0; ofm < FILTER_OFM_MAX; ofm += packed_values) {
```
Shouldn't leftovers be checked here? Looks like redundant data may be written in this loop
For example for `FILTER_OFM_NUM = 12` at the second `iter` iteration all of 8 values from `out1` and `out2` will be written, but only 4 are needed

mpenkows replied:
Right. Done.

sshlyapn commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv4_int8.cpp:
```diff
@@ -0,0 +1,113 @@
+    jit.Merge(MakeTypeJitConstants(GetAccumulatorType(params), "ACCUMULATOR"));
+    jit.Merge(MakeTypeJitConstants(GetActivationType(params), "ACTIVATION"));
+
+
```
Please delete extra line

mpenkows replied:
Done.

mpenkows commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv4_int8.cpp:
```diff
@@ -0,0 +1,113 @@
+    k.EnableSubGroup();
+    k.EnableBiasPerFeature();
+    k.EnableNonBiasTerm();
+    k.EnableBatching();
```
Batching support removed.

mpenkows commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv4_int8.h:
```diff
@@ -0,0 +1,49 @@
+
+protected:
+    WeightsLayout GetPreferredWeightsLayout(const convolution_params&) const override {
+        //return WeightsLayout::iyxo;
```
Done.

mpenkows commented:
I added a test for fusing and it has passed the CI.
